src/lib.rs

```diff
@@ -51,61 +61,77 @@ impl TriggerExecutor for CommandTrigger {
     }
 
     async fn run(self, _config: Self::RunConfig) -> Result<()> {
-        let components = self.components;
-        let engine = Arc::new(self.engine);
-
-        // TODO:
-        // This only uses first component for now.
-        let executor = Executor::new(engine, components[0].clone());
-
-        executor.run().await
+        self.handle().await
     }
+}
+
+#[async_trait]
+impl TriggerInstancePre<RuntimeData, CommandTriggerConfig> for CommandInstancePre {
+    type Instance = CommandInstance;
 
-    // TODO:
-    // This function assumes WASI P1
     async fn instantiate_pre(
-        engine: &Engine<Self::RuntimeData>,
+        engine: &Engine<RuntimeData>,
         component: &AppComponent,
-        _config: &Self::TriggerConfig,
-    ) -> Result<EitherInstancePre<Self::RuntimeData>> {
-        let module = component.load_module(engine).await?;
-        Ok(EitherInstancePre::Module(
-            engine
-                .module_instantiate_pre(&module)
-                .with_context(|| format!("Failed to instantiate component '{}'", component.id()))?,
-        ))
+        _config: &CommandTriggerConfig,
+    ) -> Result<CommandInstancePre> {
+        // Attempt to load as a component and fallback to loading a module
+        if let Ok(comp) = component.load_component(engine).await {
+            Ok(CommandInstancePre::Component(
+                engine.instantiate_pre(&comp)?,
+            ))
+        } else {
+            let module = component.load_module(engine).await?;
```

Comment:
If both of the load paths fail the error reported will be something like "failed to load or invalid module" which is pretty misleading because we swallow the `load_component` error. I think we should do something like (albeit uglier) to retain the chain of failures:

```
        match component.load_component(engine).await {
            Ok(c) => Ok(CommandInstancePre::Component(
                engine.instantiate_pre(&c)?,
            )),
            Err(component_load_err) => {
                match component.load_module(engine).await {
                    Ok(m) => Ok(CommandInstancePre::Module(
                        engine.module_instantiate_pre(&m)?,
                    )),
                    Err(module_load_err) => {
                        Err(anyhow!("{module_load_err}")
                            .context(component_load_err)
                            .context("failed to load component or module"))
                    }
                }
            }
        }
```